File: library/src/main/java/com/raxdenstudios/square/SquareFragment.java
package com.raxdenstudios.square;

import android.app.Activity;
import android.app.Fragment;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.raxdenstudios.square.interceptor.type.FragmentInterceptor;
import com.raxdenstudios.square.interceptor.manager.FragmentInterceptorManager;
import com.raxdenstudios.square.interceptor.manager.InterceptorManagerFactory;

import java.util.List;

/**
 * Created by Ángel Gómez on 18/12/2016.
 */

public abstract class SquareFragment extends Fragment {

    private FragmentInterceptorManager mInterceptorManager;

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        getInterceptorManager().onSaveInstanceState(outState);
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        getInterceptorManager().onAttach(activity);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        getInterceptorManager().onAttach(context);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getInterceptorManager().onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return getInterceptorManager().onCreateView(inflater, null, container, savedInstanceState);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getInterceptorManager().onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getInterceptorManager().onActivityCreated(savedInstanceState);
    }

    @Override
    public void onStart() {
        super.onStart();
        getInterceptorManager().onStart();
    }

    @Override
    public void onResume() {
        super.onResume();
        getInterceptorManager().onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        getInterceptorManager().onPause();
    }

    @Override
    public void onStop() {
        super.onStop();
        getInterceptorManager().onStop();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        getInterceptorManager().onDestroyView();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        getInterceptorManager().onDestroy();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        getInterceptorManager().onDetach();
    }

    /* Callbacks */

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        getInterceptorManager().onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        getInterceptorManager().onConfigurationChanged(newConfig);
    }

    /* Support methods */

    private FragmentInterceptorManager getInterceptorManager() {
        if (mInterceptorManager == null) {
            mInterceptorManager = (FragmentInterceptorManager) InterceptorManagerFactory.buildManager(this);
            addCustomInterceptorToStack(mInterceptorManager.getInterceptors());
        }
        return mInterceptorManager;
    }

    protected void addCustomInterceptorToStack(List<FragmentInterceptor> interceptors) {

    }

}

File: square/src/main/java/com/raxdenstudios/square/SquareFragment.kt
package com.raxdenstudios.square

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.raxdenstudios.square.interceptor.FragmentInterceptor
import com.raxdenstudios.square.manager.FragmentInterceptorManager
import com.raxdenstudios.square.manager.InterceptorManagerFactory

/**
 * Created by Ángel Gómez
 *
 * SquareFragment is an abstract class that adds interceptor functionality to the Fragment.
 */
abstract class SquareFragment : Fragment() {

    private var fragmentInterceptorManager: FragmentInterceptorManager? = null

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        interceptorManager.onActivityResult(requestCode, resultCode, data)
    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)
        interceptorManager.onConfigurationChanged(newConfig)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        interceptorManager.onSaveInstanceState(outState)
    }

    override fun onAttach(activity: Activity?) {
        super.onAttach(activity)
        interceptorManager.onAttach(activity)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        interceptorManager.onAttach(context)
    }

    override fun onAttachFragment(childFragment: Fragment?) {
        super.onAttachFragment(childFragment)
        interceptorManager.onAttachFragment(childFragment)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        interceptorManager.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val parentView = super.onCreateView(inflater, container, savedInstanceState)
        return interceptorManager.onCreateView(inflater, parentView, container, savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        interceptorManager.onViewCreated(view, savedInstanceState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        interceptorManager.onActivityCreated(savedInstanceState)
    }

    override fun onStart() {
        super.onStart()
        interceptorManager.onStart()
    }

    override fun onResume() {
        super.onResume()
        interceptorManager.onResume()
    }

    override fun onPause() {
        super.onPause()
        interceptorManager.onPause()
    }

    override fun onStop() {
        super.onStop()
        interceptorManager.onStop()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        interceptorManager.onDestroyView()
    }

    override fun onDestroy() {
        super.onDestroy()
        interceptorManager.onDestroy()
    }

    override fun onDetach() {
        super.onDetach()
        interceptorManager.onDetach()
    }

    /* Support methods */

    protected abstract fun setupInterceptors(interceptorList: List<FragmentInterceptor<*>>)

    private val interceptorManager: FragmentInterceptorManager
        get() {
            return fragmentInterceptorManager ?: initInterceptorManager()
        }

    private fun initInterceptorManager(): FragmentInterceptorManager {
        fragmentInterceptorManager = InterceptorManagerFactory.buildManager(this) as FragmentInterceptorManager
        fragmentInterceptorManager?.apply {
            val interceptorList = mutableListOf<FragmentInterceptor<*>>()
            setupInterceptors(interceptorList)
            interceptorList.forEach { addInterceptor(it) }
        }
        return fragmentInterceptorManager!!
    }

}
